pkg/controllers/cspc-controller/sync.go

```diff
@@ -36,6 +38,17 @@ import (
 	"k8s.io/klog"
 )
 
+type upgradeParams struct {
+	cspc   *cstor.CStorPoolCluster
+	client clientset.Interface
+}
+
+type upgradeFunc func(u *upgradeParams) (*cstor.CStorPoolCluster, error)
```

Comment:
Will this make etcd calls to update CSPC? If yes just thinking why can't we make this internal Upgrade as an atomic operation? Is there any reason why we not selected?

Reply:
Some times the operations may not be atomic or we have steps that need to be performed in some order. If by atomic you mean single operation then it is safer to have a framework that allows you to have both options of having separate or a single operation. We can always write functions that don't make any etcd calls if not required but having the option will be better for any future requirements.